### ISSUE TITLE
Remove the canvas renderer addon

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#4779]
Context from https://github.com/xtermjs/xterm.js/issues/4737#issuecomment-1712512066

> some history on that is that years ago, back when the canvas renderer was built in to the core, we wanted to remove it completely to reduce maintenance overhead. We ended up keeping it pretty much exclusively for the cases where the webgl renderer would not work like some Linux setups, VMs, iOS/Safari. The case of Safari not supporting webgl2 was the biggest of these which made us instead split canvas into an addon which reduces code loading unless the user opts in or falls back to canvas.
> 
> But since https://github.com/xtermjs/xterm.js/pull/4605 and its follow ups, plus the fact that webgl2 has shipped in Safari/WebKit, I think that covers all our bases. So we could rip out the ~1-2k lines of code and also move the shared renderer code in core into the addon (which probably won't reduce core bundle size but it will simplify things).

We should remove the canvas renderer in the next major version to reduce maintenance overhead.

**Comment by jerch** [comment]
Oh dear - it happened. RIP dear canvas renderer...

**Comment by Tyriar** [comment]
@davidfiala the dom renderer is actually way faster now than it used to be. That's why removing it was being considered again to begin with.

The issue with keeping it around is that I end up having to waste time maintaining it and any feature implemented needs to be considered for 3 different renderers. It also had some serious design flaws, such as cells not being able to draw beyond their line which means for example that underscores could become completely invisible. 

Additionally I'd like to eventually port the webgl one to webgpu but definitely wasn't going to attempt that only to them maintain 4 implementations.

For the context loss issue there are hooks to listen for that and handle it, you shouldn't be facing data/connection loss as it's just the renderer that needs to be switched. I'd like this to be handled as smoothly as possible so if you have ideas to make it better I'd love to hear them. VS Code has these same gl context limitations and I haven't heard any complaints for a long time after the initial fixes to context loss handling.

**Comment by jerch** [comment]
> Back to your original request for ideas: Here's a brainstorming question.... Could xtermjs be aware of multiple instances of itself, and multiplex a single GL context to render all live instances? ie, with 1 context, write N different terminals to N different output locations? ;) Sorry if this is an uninformed idea... I'm not very familiar with webgl.

Hmm I wasn't aware, that browser engines restrict gl contex usage to those low numbers  (btw they also do this for audio contex, which for some reason I was aware of...) To me your idea seems to be the only "workaround" in the long run.

But note that this will be rather cumbersome to achieve, as it would involve a global (as of js-context-wide) driver-like manager dealing with the scarce gl resource. xterm.js makes heavy usage of texture RAM from the glyph cache, which ofc will see even worse thrashing behavior, when shared across several terminals at once. (Idk how browser engines deal with that on the level of separate gl contexes, whether they have access to faster blitting primitives than webgl offers)

**Comment by Tyriar** [comment]
> Does DOM renderer add scroll history on-demand, or leave it permanently in the DOM? (to be fair, it sounds hard to accomplish on-demand history loading/unloading)

@davidfiala it doesn't, it should perform the same with 0 and 100000 scrollback as it only considers the viewport. Unless you have `screenReaderMode` on in which case performance would tank regardless of which renderer you're using.

> In vscode: I loaded up a bunch of terminals all running a constantly-changing visual app. In this case cmatrix. Once you hit a certain number of terminals, old onest ones die out as you open new ones:

This is just a case that people hit so rarely that no one reports it. I think we handle recovering if the context was lost when it was in the background and then replace it. For this particular case we could handle this such that if there are so many visible at once that there are no contexts available, for that particular terminal temporarily fallback to DOM.

> Back to your original request for ideas: Here's a brainstorming question.... Could xtermjs be aware of multiple instances of itself, and multiplex a single GL context to render all live instances? ie, with 1 context, write N different terminals to N different output locations? ;) Sorry if this is an uninformed idea... I'm not very familiar with webgl.

I'm actually working on a webgpu renderer for the monaco editor and I actually had this exact idea recently to share a canvas between multiple editors in order to solve the problem of notebooks that can have an unbounded number of editor instances. The main difficulty for applying this idea to xterm.js are:

- It might be difficult to support due to assumptions of it having full control of the canvas
- How to craft the API such that it shares the canvas. Maybe something like this? `new WebglAddon({ canvas: { x, y, w, h } })`

I wouldn't be able to spend time on this but someone else could have a look if it's very important to them. Personally I think for xterm.js this is such an edge case that it's not worth it when you can just fallback. Also I'm not sure if webgpu has these same limitations, porting it to that could be the most forward-looking solution.

> Could we learn here from other highly integrating 3d libs like three.js? Do they have a global gl manager?

@jerch I haven't used it but I had a look at their docs and I don't think they have anything special here.

**Comment by jerch** [comment]
@Tyriar I know this is like a never ending story - is it about time to re-eval a possible usage of offscreencanvas here? It seems, that the feature coverage got much better recently, also see the support table here: https://developer.mozilla.org/en-US/docs/Web/API/OffscreenCanvas#browser_compatibility

**Comment by Tyriar** [comment]
@jerch that would only be useful if we were going to do rendering in a worker which isn't worth the hassle imo as it would cause sync problems issues and add a bunch of complexity when most draws are already < 1ms CPU time. I think offloading the parsing if anything would be the way to go, but I've looked at that several times and it's difficult without encoding all our state into array buffers.

We do use OffscreenCanvas to measure text metrics now:

https://github.com/Tyriar/xterm.js/blob/8f8c94a73405f742021098d485bd689451e8818f/src/browser/services/CharSizeService.ts#L104-L127

The main reason we would use it beyond that is for the texture atlas and I think all that would give is just make this a little more elegant:

https://github.com/Tyriar/xterm.js/blob/8f8c94a73405f742021098d485bd689451e8818f/src/browser/renderer/shared/TextureAtlas.ts#L1095-L1100

**Comment by jerch** [comment]
@davidfiala Have you actually benchmarked the DOM renderer of a more recent version? Here are some numbers from my side: https://github.com/xtermjs/xterm.js/pull/4605#issuecomment-1656126195

Plz note the comparison with the canvas renderer, which actually performs worse with its 100% CPU saturation :smiley_cat:

**Comment by Tyriar** [comment]
> DOM renderer meets 1 (fails 2,3)

It's plenty fast for the vast majority of use cases. Here's an example of the sort of times I get on my machine for a full viewport render (189x26 terminal running `tree` on Windows):

![image](https://github.com/user-attachments/assets/1b35a200-12aa-48cf-aedd-58b8d0e846ef)

Compared to webgl:

![image](https://github.com/user-attachments/assets/6893cb63-0945-41f7-9205-9a4bb84fd404)

This is perfectly reasonable and reliable now (it wasn't before @jerch looked into it). The main thing you're missing out on with DOM is custom glyphs and render times/input latency as low as we can get them.

> Canvas renderer meets 1 and 2 (not so much 3)

I don't think the canvas renderer meets 2, on some setups like when GPU acceleration is not enabled on the machine it will render using the CPU and become extremely slow. VS Code had some special frame measuring logic that I was very glad to get rid of. Here was the fallback chain:

1. Try webgl renderer
2. If it throws, try the canvas renderer and measure x frames to ensure they are reasonable
3. If it doesn't meet the requirements fallback to dom

> Maybe DOM renderer should be the one we pick on, if we really want to drop some eng load?
> ...
> Should we stack rank our renderers and cut the bottom? Why should DOM renderer stay while canvas goes? My ranking places canvas above DOM. 

There must always be a fallback to DOM because the bad number canvas showed above, in the end it's rendering with the GPU anyway in the browser engine, just with a bunch of layout stuff happening (which we now minimize).

---

I appreciate the passion, but I think you should re-evaluate the DOM renderer which really is great now and in many cases better than canvas. We removed canvas because it's obsolete and has become a significant burden, it only exists to begin with because I didn't know webgl when I implemented [7 years ago](https://code.visualstudio.com/blogs/2017/10/03/terminal-renderer). The canvas renderer also missed out on some advanced features we recently did because it's too much of a hassle to implement (see skipped tests).

**Comment by Tyriar** [comment]
> BTW, is the source tree known reasonable at the time that canvas was removed? I may need to snapshot or keep a fork with it cherry pick'd back in for a while if I intend to adopt any other improvements that v6 will include.

You'd be best sticking with the release before https://github.com/xtermjs/xterm.js/pull/5092 as this and after contained some really big changes which haven't been battle tested yet (only in VS Code insiders, not stable). Unfortunately we don't have a good link between which commit points at which release, but basically it's the first one that does not include any .mjs files.

> Meanwhile, I hope that the links https://github.com/xtermjs/xterm.js/issues/4779#issuecomment-2231514623 in can help with a path forward to reduce WebGL context usage, at a minimum. It doesn't solve my argument that we cannot trust in webgl, but it does at least provide 16+ contexts and maybe reduce the chance of a loss.

I think canvas sharing would be a good thing to support but not sure I'll have time to work on it, though I may experiment with the idea in the monaco editor codebase.

> Far more people are using dev tools to manage production than they should be... I doubt we'll ever hear about bug reports though. People may not even realize that xtermjs powers vscode.

I triage all the issues coming into the vscode repo too so I have a pretty good idea of the stability of it. I'm actually a little surprised how little I've heard post it's removal was announced in https://code.visualstudio.com/updates/v1_90#_-removal-of-the-canvas-renderer

> I know the upper-bound cost of using it isn't prohibitive and it won't ever freak out the DOM tree because it never changes the DOM tree (I think?)

Not sure what you mean here, a lot of elements are recycled and reused but there's a lot of changes inside the `.xterm-rows` element.

> fallback to DOM in my experience was pretty hard to code for.

If you have suggestions to make this better let us know. I know it's not great as I see it's not even properly handled in VS Code yet. Pretty low priority on my end though,.

> Dropping keystrokes

They're completely separate systems, you should never lose any keystrokes as input is routed elsewhere

**Comment by jerch** [comment]
@davidfiala If your software uses a recent version of xterm.js - could you plz check first, if the DOM renderer is really that bad for you? This should be quite easy, just dont load the canvas and webgl addons.

We are not getting anywhere, if we keep telling you, that the DOM renderer got much better in the last months, and you respond to that with vaguely phrased issues of a much older version, without having tested against a newer version. I dont say, that it will work flawless now (in fact - we never had a single flawless renderer to begin with, they all have their rough edges for very different reasons), but perf got so much better, that your point 2 should hardly be an issue anymore.

**Comment by Tyriar** [comment]
@davidfiala we don't use workers right now but it we were to I doubt we would move the rendering there as parsing incoming data is the more expensive task. There's also the concern that a worker could increase input latency which is one of the more important perf metrics.

**Comment by jerch** [comment]
Regarding webworkers (somewhat offtopic):
Initially I went with a worker approach in the image addon basically following the idea "yep, lots of data to digest here, which would block the main thread unnecessarily long". This was true as long as I did the image decoding in JS, but with the switch to WASM it was actually much worse due to the needed messaging overhead + state synchronization. So quite mixed/poor results with high IO pressure and workers. To me it seems, that the standard worker interface is too fat for serious high IO perf stuff.

This would be much better with SABs for data transfer, but those are hard to set up correctly due security constraints (geez, for quite some time it was not clear, if they would come back at all). And beside the bytearray serialization they also need a proper locking strategy, so the impl bar is quite high. All in all - workers are not that sexy for typical xterm.js load (at least for me not anymore). Geez, I wish we had an Erlang clone in browsers, or something similar with lightweight threading...
